mesonbuild/compilers/mixins/clang.py

```diff
@@ -101,13 +101,24 @@ def has_function(self, funcname: str, prefix: str, env: 'Environment', *,
                      dependencies: T.Optional[T.List['Dependency']] = None) -> T.Tuple[bool, bool]:
         if extra_args is None:
             extra_args = []
-        # Starting with XCode 8, we need to pass this to force linker
-        # visibility to obey OS X/iOS/tvOS minimum version targets with
-        # -mmacosx-version-min, -miphoneos-version-min, -mtvos-version-min etc.
-        # https://github.com/Homebrew/homebrew-core/issues/3727
         # TODO: this really should be communicated by the linker
-        if isinstance(self.linker, AppleDynamicLinker) and mesonlib.version_compare(self.version, '>=8.0'):
-            extra_args.append('-Wl,-no_weak_imports')
+        if isinstance(self.linker, AppleDynamicLinker):
+            # Starting with XCode 8, we need to pass this to force linker
+            # visibility to obey OS X/iOS/tvOS minimum version targets with
+            # -mmacosx-version-min, -miphoneos-version-min, -mtvos-version-min etc.
+            # https://github.com/Homebrew/homebrew-core/issues/3727
+            if mesonlib.version_compare(self.version, '>=8.0'):
+                extra_args.append('-Wl,-no_weak_imports')
+            # XCode 9 introduced two new warnings:
+            # -Wunguarded-availability and Wunguarded-availability-new
+            # The former is off by default, the latter on by default,
+            # -Wunguarded-availability-new only warns about APIs introduced
+            # in or after macOS 10.13, iOS 11, tvOS 11 or watchOS 4.
+            # Since macOS 10.13 was EOLed on 1 Dec 2020, there's no need to
+            # increase the severity level of the former warning.
+            # https://github.com/lovell/sharp/issues/3438
+            if mesonlib.version_compare(self.version, '>=9.0'):
+                extra_args.append('-Werror=unguarded-availability-new')
```

Comment:
I think while we're here it cannot hurt to increase the severity level of checks for people trying to build code to deploy on a relatively recent OS.

I mean, if we were going to drop support for old OSes we'd presumably drop `-Wl,-no_weak_imports`, right?

Reply:
Indeed, but `-Wunguarded-availability-new` would probably also warn about using APIs introduced in macOS 12, if you compile with the `MACOSX_DEPLOYMENT_TARGET="11.0"` env set, for example.

Anyways, I think this commit can be reverted, it looks like `-Wl,-no_weak_imports` would already catch this (if you use this in combination with `prefix: '#include <...>'`).

Reply:
Ah, so this is basically just several different flags that all kind of do the same thing?

...

If there's some flag that helps out such that even function tests that forgot to add a prefixed `#include` are correctly detected, that would be quite cool. Do the new flags do that, at least?

Reply:
> Ah, so this is basically just several different flags that all kind of do the same thing?

I think so, which is somewhat confusing, but I'm not a macOS expert. :sweat_smile: 

> If there's some flag that helps out such that even function tests that forgot to add a prefixed `#include` are correctly detected, that would be quite cool. Do the new flags do that, at least?

Indeed, that would fix this. But unfortunately, I don't know of any compile/link flags that can do this.

Reply:
> If there's some flag that helps out such that even function tests that forgot to add a prefixed #include are correctly detected, that would be quite cool. Do the new flags do that, at least?

There is no way to do this. That's why I added the note in the docs to discourage using the check without the right `#include` in the prefix.

Reply:
> There is no way to do this. That's why I added the note in the docs to discourage using the check without the right `#include` in the prefix.

Thanks for confirming and mentioning this in the docs!

So, to summarize:
1. Meson cannot do anything about this in terms of implementation.
2. Meson already recommends using the `prefix: '#include <...>'` argument in the docs, which would fix such issues.
3. Meson could issue a warning if that argument is not used (tracked in issue https://github.com/mesonbuild/meson/issues/3482).
4. For macOS, Meson could recommend the `-Werror=unguarded-availability-new` compilation flag in the docs to detect such issues at a later stage.
5. This PR is only useful as a regression test to check if the `MACOSX_DEPLOYMENT_TARGET` env is honored.

Since (5) is actually something that should be handled by the compilers (and not the build system), my feeling is that I should abandon this PR. Am I right?

Reply:
I see, abandoning this PR. Thanks for allowing me to test this on CI. 👍